components/brave_vpn/browser/connection/wireguard/brave_vpn_wireguard_connection_api_base.cc

```diff
@@ -0,0 +1,143 @@
+
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_vpn/browser/connection/wireguard/brave_vpn_wireguard_connection_api_base.h"
+
+#include "brave/components/brave_vpn/common/brave_vpn_utils.h"
+#include "brave/components/brave_vpn/common/mojom/brave_vpn.mojom.h"
+#include "brave/components/brave_vpn/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+
+namespace brave_vpn {
+
+using ConnectionState = mojom::ConnectionState;
+
+BraveVPNWireguardConnectionAPIBase::BraveVPNWireguardConnectionAPIBase(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    PrefService* local_prefs)
+    : BraveVPNOSConnectionAPI(url_loader_factory, local_prefs) {}
+
+BraveVPNWireguardConnectionAPIBase::~BraveVPNWireguardConnectionAPIBase() =
+    default;
+
+void BraveVPNWireguardConnectionAPIBase::SetSelectedRegion(
+    const std::string& name) {
+  GetRegionDataManager().SetSelectedRegion(name);
+}
+
+void BraveVPNWireguardConnectionAPIBase::OnWireguardServiceLaunched(
+    bool success) {
+  UpdateAndNotifyConnectionStateChange(
+      success ? ConnectionState::CONNECTED : ConnectionState::CONNECT_FAILED);
+}
+
+void BraveVPNWireguardConnectionAPIBase::OnWireguardKeypairGenerated(
+    brave_vpn::wireguard::WireguardKeyPair key_pair) {
+  if (!key_pair.has_value()) {
+    VLOG(1) << __func__ << " : failed to get keypair";
+    UpdateAndNotifyConnectionStateChange(ConnectionState::CONNECT_FAILED);
+    return;
+  }
+  const auto [public_key, private_key] = key_pair.value();
+  GetAPIRequest()->GetWireguardProfileCredentials(
+      base::BindOnce(
+          &BraveVPNWireguardConnectionAPIBase::OnGetProfileCredentials,
+          base::Unretained(this), private_key),
+      GetSubscriberCredential(local_prefs()), public_key, GetHostname());
+}
+
+void BraveVPNWireguardConnectionAPIBase::Connect() {
+  if (GetConnectionState() == ConnectionState::CONNECTED) {
+    Disconnect();
+    return;
+  }
+
+  VLOG(2) << __func__ << " : start connecting!";
+  SetLastConnectionError(std::string());
+  UpdateAndNotifyConnectionStateChange(ConnectionState::CONNECTING);
+
```

Comment:
Q) if we can get previously used creds from `kBraveVPNWireguardProfileCredentials`, could we call `PlatformConnectImpl()` directly here?

Reply:
Good catch, thanks for checking. Added a check for existing hostname and call to connect if we have credentials. In RAS we reset connection info if connect was failed, here I implemented similar in last commits to subscribe on connect_failed state and reset saved info to fetch region/profile creds again next time